rfd/0181-pg-database-access-webui.md

```diff
@@ -0,0 +1,442 @@
+
+In addition to executing queries, the interactive shell will implement some
+backlash (\\) commands from `psql`. Users can fetch the list of supported
+commands by calling the help command (`\?` or `help`):
+
+```shell
+postgres=> \?
+General:
+  \q          Terminate the session.
+  \teleport   Show Teleport interactive shell information, such as execution limitations.
+
+Informational:
+  \d              List tables, views, and sequences.
+
+Connection/Session:
+  \session   Display information about the current session, like user, roles, and database instance.
+```
+
+Supported list of commands and output examples:
+
+* `\q`: No output. Terminates the session.
+
+* `\teleport`: Presents the shell information.
+  ```
+  postgres=> \teleport
+  Teleport PostgreSQL interactive shell (v16.2.0-dev)
+  ```
+
+* `\d`: Describes the tables, views and sequences.
+  ```
+  Schema Name             Type     Owner    
+  ------ ---------------- -------- -------- 
+  public employees        table    alice    
+  public employees_id_seq sequence alice    
+  public event            table    postgres 
+  public events           table    postgres 
+  ```
+
+* `\session`: Presents session information.
+  ```
+  Connected to "pg" instance as "alice" user.
+  ```
+
+
+##### Limitations/Unpported commands
+
+Teleport PostgreSQL interactive shell will not be a complete feature pair with
+`psql`. Those limitations will be due to security measures or the shell's
+simplicity. Given those limitations, messages will be shown to the users,
+displaying a description and direction on executing the desired command
+(if applicable).
+
+Unsupported backslash (\\) commands from psql will only display a failure
+message:
+
+```shell
+postgres=> \set a "hello"
+Invalid command \set.
+Try "help" or "\?" for the list of supported commands.
+
+postgres=>
+```
+
+Other limitations, such as query size limit, will display a more complete message:
+
+```shell
+postgres=> INSERT INTO ... # Long query.
+ERROR: Unable to execute query. Max query size limit (2048 characters) exceeded.
+For long queries, execute it using `tsh db` commands.
+
+postgres=>
+```
+
+### Product usage
+
+To track PostgreSQL databases access through Web UI, we need to differentiate
+them from direct sessions. To do so, we'll add a new field to the
+[`SessionStartDatabaseMetadata`](https://github.com/gravitational/teleport/blob/5e921b33b9173ccde8fe321d35534b04ebbd5d93/proto/prehog/v1alpha/teleport.proto#L223)
+event to include this information:
+
+```diff
+// SessionStartDatabaseMetadata contains additional information about database session.
+message SessionStartDatabaseMetadata {
+  // database type.
+  string db_type = 1;
+  // database protocol.
+  string db_protocol = 2;
+  // database origin source.
+  string db_origin = 3;
++ // access_through describes whether the database was accessed by speaking to
++ // the proxy service or using Web UI interactive shell.
++ string access_through = 4;
```

Comment:
LGTM. just curious how would the agent tell if the connection is from web ui ? by session id?

Reply:
We'll need to propagate the session ID information from the proxy to ensure the ID on the Web UI matches the session recording. When this is done, we can assume that the session is from web UI, or we could add a new field to indicate this (which I think is a better approach).

Reply:
In addition to `access_through`, maybe we can start reporting `user-agent` retrieved from the protocol level? Like `application_name` from postgres client? https://postgresqlco.nf/doc/en/param/application_name/. This could be beneficial outside this feature as well. Could be a cool hackathon project too xD